src/QsCompiler/CompilationManager/CompilationUnit.cs

```diff
@@ -483,25 +483,29 @@ private QsCallable GetImportedCallable(CallableDeclarationHeader header)
         {
             // TODO: this needs to be adapted if we want to support external specializations
             if (header == null) throw new ArgumentNullException(nameof(header));
-            var importedSpecs = this.GlobalSymbols.ImportedSpecializations(header.QualifiedName);
             if (Namespace.IsDeclarationAccessible(false, header.Modifiers.Access))
             {
-                var definedSpecs = this.GlobalSymbols.DefinedSpecializations(header.QualifiedName);
+                var definedSpecs = GlobalSymbols.DefinedSpecializations(header.QualifiedName);
                 QsCompilerError.Verify(definedSpecs.Length == 0,
                                        "external specializations are currently not supported");
             }
 
-            var specializations = importedSpecs.Select(imported =>
-            {
-                var (specHeader, implementation) = imported;
-                var specSignature = specHeader.Kind.IsQsControlled || specHeader.Kind.IsQsControlledAdjoint 
-                    ? SyntaxGenerator.BuildControlled(header.Signature) 
-                    : header.Signature;
-                return new QsSpecialization(specHeader.Kind, header.QualifiedName, specHeader.Attributes,
-                    specHeader.SourceFile, specHeader.Location, specHeader.TypeArguments, specSignature,
-                    implementation, specHeader.Documentation, QsComments.Empty);
-            })
-            .ToImmutableArray();
+            var specializations =
+                GlobalSymbols
+                .ImportedSpecializations(header.QualifiedName)
+                .Where(specialization => specialization.Item1.SourceFile.Equals(header.SourceFile))
```

Comment:
Why are you filtering by source file? I realize that we currently only support that scenario, but I'd rather not.

Reply:
I updated the condition to filter by source only if the callable is internal. Even if we support external specializations, specializations for an internal callable can only be defined within the same assembly, because it's not visible from anywhere else. So if I don't filter by source, a callable could get specializations that don't belong to it - does that make sense?

Reply:
Ah, I see. Yes, that makes sense, thanks!